scrape/scrape.go

```diff
@@ -714,19 +715,29 @@ type scrapeCache struct {
 	// seriesCur and seriesPrev store the labels of series that were seen
 	// in the current and previous scrape.
 	// We hold two maps and swap them out to save allocations.
-	seriesCur  map[uint64]labels.Labels
-	seriesPrev map[uint64]labels.Labels
+	seriesCur  map[uint64]seriesEntry
+	seriesPrev map[uint64]seriesEntry
 
 	metaMtx  sync.Mutex
 	metadata map[string]*metaEntry
 }
 
+// seriesEntry holds meta information about an active series.
+type seriesEntry struct {
+	labels labels.Labels
+	meta   storage.Metadata
+}
+
 // metaEntry holds meta information about a metric.
 type metaEntry struct {
-	lastIter uint64 // Last scrape iteration the entry was observed at.
-	typ      textparse.MetricType
-	help     string
-	unit     string
+	lastIter             uint64 // Last scrape iteration the entry was observed at.
+	lastIterObservedType uint64 // Last scrape iteration the type was observed at.
```

Comment:
Currently metadata is not actually zeroed when there is new metadata per scrape, each field (type/unit/help) are all individually are updated if and only if they are seen, all of which in their setters bump the `lastIter` field to current iteration. Since we want the effect of setting any of these to "zero" out the metadata when we next inform the TSDB of a metadata update, we need to track which we actually updated this scrape and treat the other fields as zeroed.

The complexity with actually zeroing out the other fields when any of type/unit/help is set the first time during a scrape is that the lock is only held for the duration of setting the single metadata field (type or unit or help) so zeroing the other fields at that time will mean that the metadata goes missing temporarily for the field you didn't update (i.e. you set type and clear unit/help since it's the first update for this "iter", once you release the lock now the old values of unit/help aren't there anymore even though the very next line of the scrape you parse could be restoring the unit/help values to what they were just before you cleared them if they aren't being zeroed this scrape).

So for consistency what I was doing here is keeping the current behavior of "don't zero the other metadata so that metadata constantly moves only forwards and is never reset" for the current metadata consumers, but adding enough information to the variables to know actually which fields were present when we did observe a change and using that information to properly make those fields zeroed when passing along metadata to the TSDB and only including the metadata fields that were set with the scrape iteration that the metadata changed.